include/picongpu/particles/collision/detail/ListEntry.hpp

```diff
@@ -0,0 +1,292 @@
+                                }
+                            }
+                            if(ptrToIndicies == nullptr)
+                            {
+#if BOOST_COMP_HIP
+                                // do nothing
+#elif BOOST_LANG_CUDA
+                                __trap();
+#else
+                                throw std::runtime_error("Out of memory in collisions.");
+#endif
+                            }
+                        }
+                        // reset counter
+                        size = 0u;
+                    }
+
+                    /* Release allocated heap memory.
+                     *
+                     * @param acc alpaka accelerator
+                     * @param deviceHeapHandle Heap handle used for allocating storage on device.
+                     */
+                    template<typename T_acc, typename T_DeviceHeapHandle>
+                    DINLINE void finalize(T_acc const& acc, T_DeviceHeapHandle& deviceHeapHandle)
+                    {
+                        if(ptrToIndicies != nullptr)
+                        {
+#if(PMACC_CUDA_ENABLED == 1)
+                            deviceHeapHandle.free(acc, (void*) ptrToIndicies);
+                            ptrToIndicies = nullptr;
+#else
+                            delete(ptrToIndicies);
+#endif
+                        }
+                    }
+
+
+                    /* Shuffle list entries.
+                     *
+                     * @param acc alpaka accelerator
+                     * @param rngHandle random number generator handle
+                     */
+                    template<typename T_Acc, typename T_RngHandle>
+                    DINLINE void shuffle(T_Acc const& acc, T_RngHandle& rngHandle)
+                    {
+                        using UniformUint32_t = pmacc::random::distributions::Uniform<uint32_t>;
+                        auto rng = rngHandle.template applyDistribution<UniformUint32_t>();
+                        // shuffle the particle lookup table
+                        for(uint32_t i = size; i > 1; --i)
+                        {
+                            /* modulo is not perfect but okish,
+                             * because of the loop head mod zero is not possible
+                             */
+                            int p = rng(acc) % i;
+                            if(i - 1 != p)
+                                swap(ptrToIndicies[i - 1], ptrToIndicies[p]);
+                        }
+                    }
+
+
+                private:
+                    /* Swap two list entries.
+                     *
+                     * @param v0 index of the 1st entry to swap
+                     * @param v0 index of the 2nd entry to swap
+                     */
+                    DINLINE void swap(uint32_t& v0, uint32_t& v1)
+                    {
+                        uint32_t tmp = v0;
+                        v0 = v1;
+                        v1 = tmp;
+                    }
```

Comment:
I think ideally this function should probably not be here, but as a some generic pmacc utility function like `std::swap<T>`. But it's okay as is now.

Reply:
`std::swap` should not work on GPU it is not a constexpr function. This will be changed with C++20.